kallsyms/kallsyms.go

```diff
@@ -0,0 +1,468 @@
+func (m *Module) finish() {
+	sort.Slice(m.symbols, func(i, j int) bool {
+		return m.symbols[i].offset >= m.symbols[j].offset
+	})
+
+	// Synthesize fileID if it was not available via /sys
+	if m.fileID.Compare(libpf.FileID{}) == 0 && len(m.symbols) > 0 {
+		// Hash exports and their normalized addresses.
+		h := fnv.New128a()
+
+		h.Write(m.bytesAt(0)) // module name
+		size := uint64(m.end - m.start)
+		h.Write(libpf.SliceFrom(&size))
+
+		for _, sym := range m.symbols {
+			h.Write(m.bytesAt(sym.index))
+			addr := uint64(sym.offset)
+			h.Write(libpf.SliceFrom(&addr))
+		}
+
+		var hash [16]byte
+		fileID, err := libpf.FileIDFromBytes(h.Sum(hash[:0]))
+		if err != nil {
+			panic("calcFallbackModuleID file ID construction is broken")
+		}
+
+		log.Debugf("Fallback module ID for module %s is '%s' (num syms: %d)",
+			m.Name(), fileID.Base64(), len(m.symbols))
+	}
+}
+
+func (m *Module) Name() string {
+	return m.stringAt(0)
+}
+
+func (m *Module) Start() libpf.Address {
+	return m.start
+}
+
+func (m *Module) End() libpf.Address {
+	return m.end
+}
+
+func (m *Module) BuildID() string {
+	return m.buildID
+}
+
+func (m *Module) FileID() libpf.FileID {
+	return m.fileID
+}
+
+// LookupSymbolByAddress resolves the `pc` address to the function and offset from it.
+// On error, an empty string with zero offset is returned.
+func (m *Module) LookupSymbolByAddress(pc libpf.Address) (funcName string, offset uint) {
+	pcOffs := uint32(pc - m.start)
+	symIdx := sort.Search(len(m.symbols), func(i int) bool {
+		return pcOffs >= m.symbols[i].offset
+	})
+	if symIdx >= len(m.symbols) {
+		return "", 0
+	}
+	sym := &m.symbols[symIdx]
+	symName := m.stringAt(sym.index)
+	return symName, uint(pcOffs - sym.offset)
+}
+
+// LookupSymbol finds a symbol with 'name' from the Module.
+func (m *Module) LookupSymbol(name string) (libpf.Address, error) {
+	for _, sym := range m.symbols {
+		if m.stringAt(sym.index) == name {
+			return m.start + libpf.Address(sym.offset), nil
+		}
+	}
+	return 0, ErrNoSymbol
+}
+
+// LookupSymbolsByPrefix finds all symbols with the given prefix in from the Module.
+func (m *Module) LookupSymbolsByPrefix(prefix string) []*libpf.Symbol {
+	res := make([]*libpf.Symbol, 0, 8)
+	for _, sym := range m.symbols {
+		symName := m.stringAt(sym.index)
+		if strings.HasPrefix(symName, prefix) {
+			symAddr := m.start + libpf.Address(sym.offset)
+			res = append(res, &libpf.Symbol{
+				Name:    libpf.SymbolName(symName),
+				Address: libpf.SymbolValue(symAddr),
+			})
+		}
+	}
+	return res
+}
+
+// updateSymbolsFrom parses /proc/kallsyms format data from the reader 'r'.
+// If possible the data from previous reads is re-used to avoid allocations.
+// The Symbolizer internal state is update only if the input data is parsed
+// successfully.
+func (s *Symbolizer) updateSymbolsFrom(r io.Reader) error {
+	var mod *Module
+	var curName string
+	var syms []symbol
+	var names []byte
+
+	s.valid.Store(true)
```

Comment:
I think, we should move `s.valid.Store(true)` just after `s.modules = slices.Clone(mods)`. In the scope of this function there are some error paths and if `*Symbolizer` holds some `Module`s, `(*Symbolizer) Invalidate()` is called, then setting here `valid` to `true` adds the risk of holding invalidated data if the function returns in an error path.

Reply:
It is important that `Store(true)` is called here. Otherwise there is a race condition that `Invalidate` is called form another goroutine concurrently and we miss the notification.

If the `updateSymbolsFrom` fails, the `Refresh` function (which is the sole caller) will reset valid to false state.

This said, this also means that calling any of the symbol resolving functions are not concurrent safe. `Refresh` should hold a lock in case multiple callers try to reload things concurrently. There is no code paths where this could happen currently. But perhaps this needs to be documented - or we just add the locking now in case things change?

Reply:
This mechanism has been reworked, and fully concurrent use is now allowed. Also function description updated as needed.